README
LICENCE AND COPYRIGHT
---------------------

Copyright (C) 2006 HS - Hamburger Software GmbH & Co. KG.
All rights reserved.

This module is free software; you can redistribute it and/or modify 
it under the same terms as Perl itself.

This library is distributed in the hope that it will be useful, 
but WITHOUT ANY WARRANTY; without even the implied warranty of 
<MERCHANTIBILITY or FITNESS FOR A PARTICULAR PURPOSE.



HOW TO INSTALL
--------------

To install this package, unzip and untar the file, then build the 
makefile and use make or nmake to install:

    tar xzf Config-LotusNotes-x.xx.tar.gz
    cd Config-LotusNotes-x.xx
    perl Makefile.PL
    nmake
    nmake test
    nmake install


This package also contains an ActiveState PPM package. 
To install it, run the following command in the current directory:

    ppm install Config-LotusNotes.ppd


Changes
Changelog for Config::LotusNotes
--------------------------------

0.33 (02/07/2011)
   - fixed problem locating installations in paths that contain spaces
   - fixed option parsing problem in sample script

0.32 (06/12/2008)
   - new property "install_scope" for install variants "all users" and "just for me"
   - minor documentation updates 

0.31 (02/25/2008)
   - added support for "all users" install variant
   - updated ppm package to support perl 5.10

0.30 (02/14/2007)
   - use Config::IniHash instead of Config::IniFiles to handle reading and 
     writing notes.ini files. This module can deal with malformed ini files.
   - added demo script editnotesini.pl
   - demo scripts now get installed

0.23 (10/31/2006)
   - demo/FindNotes.pl no more dies on systems with no Notes installed.
   - improved handling of debug messages for corrupt notes.ini files.
   - documentation improvements.

0.22 (10/13/2006)
   - Removed trailing backslash from notespath() for consistency with 
     datapath() attribute.
   - Config::LotusNotes no longer dies when no COM server entries are 
     found in the registry.
   - Added mock notes install files to test directory so that the modules 
     can be tested if no Notes is installed.
   - Added tests for POD coverage

0.21 (08/21/2006)
   - First public release.
   - Separated logic for locating installs from representation of a 
     configuration by introducing new module Config::LotusNotes::Configuration.
   - Several minor improvements to make the module more compatible with CPAN.

0.2.0 (08/09/2006)
   - Complete rewrite.
   - Changed namespace to Config::LotusNotes.
   - Uses Windows registry instead of lotus.ini to locate Notes.
   - Supports multiple installs.
   - Documentation translated into English.
   - Added demo script FindNotes.pl.

0.1.0 (10/28/2003)
   - First version as Lotus::Notes::Config.


blib/script/FindNotes.pl
# This script produces a table with information on all Lotus Notes
# installs found on the local computer.
# If run with a -d commandline parameter, it also produces diagnostic output.

use strict;
use warnings;
use lib "../lib";
use Config::LotusNotes;

my %options;
$options{debug} = 1 if $ARGV[0]||'' eq '-d';

my $conf = Config::LotusNotes->new(%options);
my @all_confs = $conf->all_configurations();
my $default   = eval { $conf->default_configuration() };

print "Lotus Notes installs on node ", Win32::NodeName(), ":\n";
print "-" x (length(Win32::NodeName())+30), "\n";

if (@all_confs) {
    printf "%-8s %-7s %s\n", 'version', 'type', 'path';
    foreach my $conf (sort by_version_and_type @all_confs) {
        printf "%-8s %-7s %s %s\n", 
            $conf->version, 
            $conf->is_server ? 'server' : 'client', 
            $conf->notespath,
            ($default and $conf->notespath eq $default->notespath) 
                ? '(default)'
                : ''
                ,
            ;
    }
}
else {
    print "none\n";
}


sub by_version_and_type {
       $a->version   cmp $b->version
    || $a->is_server <=> $b->is_server
}


lib/Config/LotusNotes/Configuration.pm
package Config::LotusNotes::Configuration;
use strict;
#use warnings;
use Carp;
use Config::IniHash;
use File::HomeDir;

our $VERSION = '0.33';


# constructor ----------------------------------------------------------------

sub new {
    my ($classname, %options) = @_;
    my $scope = 'just for me';
    my $path = $options{path} or die 'no Notes install path specified';
    $path =~ s/\\$//;

    # basic checks of the directory and its content
    my $notesexe = $path.'\\nlnotes.exe';
    my $notesini = $path.'\\notes.ini';
    croak "Notes install path not found: $path" unless -d $path;
    croak "No Notes binary found in $path"      unless -f $notesexe;
    if (! -f $notesini) {
        $notesini = File::HomeDir->my_data . '\Lotus\Notes\Data\notes.ini';
        croak "No notes.ini found in $path and current user's profile"  unless -f $notesini;
        $scope = 'all users';
    }

    # Create and store object for easy access to inifile values.
    # Config::IniHash lets us treat the inifile as an ordinary hash.
    my $inihash = ReadINI($notesini, (case => 'sensitive')) 
        or croak "Error parsing $notesini";
    my $self = bless {
        notespath => $path,
        notesini  => $notesini,
        inihash   => $inihash,
        scope     => $scope,
    }, $classname;
    return $self;
}


# public accessors -----------------------------------------------------------

sub notesini  { return shift->{notesini } }
sub notespath { return shift->{notespath} }
sub datapath  { return shift->get_environment_value('Directory') }

sub is_server { return shift->get_environment_value('KitType') == 2 }
sub is_client { return shift->get_environment_value('KitType') == 1 }

sub install_scope { return shift->{scope} }

sub version {
    my ($self) = @_;
    return $self->{version} if $self->{version};  # cached value

    # try to extract the version from one of these files
    my @files_with_version = qw(nsd.exe ndgts.dll ninotes.dll memcheck.exe nnntp);
    foreach my $file (@files_with_version) {
        my $filepath = $self->notespath . "\\" . $file;
        next unless -f $filepath;
        open FILE, "< $filepath" or next; # skip errors.
        while (<FILE>) {
            # we cache the result in order to avoid repeated file access.
            return $self->{version} = $2  if /(Release (\d+\.\d+(\.\d+)?[a-z]?))/
        }
        close FILE;
    }
    croak 'could not determine Notes version';
}


# public methods -------------------------------------------------------------

sub get_environment_value {
	my ($self, $key) = @_;
	return $self->{inihash}->{Notes}->{$key};
}


sub set_environment_value {
	my ($self, $key, $value) = @_;
	if (defined $value) {
        $self->{inihash}->{Notes}->{$key} = $value;
	}
	else {
        delete $self->{inihash}->{Notes}->{$key};
	}
	return WriteINI($self->{notesini}, $self->{inihash});
}


1;


=head1 NAME

Config::LotusNotes::Configuration - Represents one Lotus Notes/Domino configuration

=head1 VERSION

This documentation refers to C<Config::LotusNotes::Configuration> 0.33, 
released Feb 7, 2011.

=head1 SYNOPSIS

  $factory = Config::LotusNotes->new();

  # access default installation
  $conf = $factory->default_configuration();

  # basic information about a configuration
  print "Version: ", $conf->version, "\n";
  print "This is a server.\n" if $conf->is_server;

  # getting and setting environment values
  $data = $conf->get_environment_value('Directory');
  $conf->set_environment_value('$NotesEnvParameter', 'value');

=head1 DESCRIPTION

A C<Config::LotusNotes::Configuration> object represents the configuration
of a local Lotus Notes/Domino installation from the view of the file system.
It lets you read and modify the Lotus Notes configuration file, F<notes.ini>,
where Notes stores its environment. 
See L<Config::LotusNotes/"The Lotus Notes environment"> 
for more information on exchanging data with Lotus Notes via the Notes environment. 

C<Config::LotusNotes::Configuration> objects also give you access to some 
basic information like install paths and the Notes version number.

To create these objects, use the 
L<default_configuration()|Config::LotusNotes/item_default_configuration> and 
L<all_configurations()|Config::LotusNotes/item_default_configuration>
methods of L<Config::LotusNotes|Config::LotusNotes>.

=head1 PROPERTIES

=over 4

=item notespath();

Returns the path where the program files are stored.

=item datapath();

Returns the path where the data files are stored.

=item notesini();

Returns the full path (including file name) of the F<notes.ini> file.

=item version();

Returns the Lotus Notes version number, e.g. 7.0 or 5.0.13a.

=item is_client();

Returns true if the configuration belongs to a client (workstation) installation.

=item is_server();

Returns true if the configuration belongs to a server installation.

=item install_scope();

Returns "just for me" or "all users" depending on the chosen setup type.

=back

=head1 METHODS

=over 4

=item new(path => $path);

Constructor, returns a C<Config::LotusNotes::Configuration> object 
representing the installation at the specified path. 

The recommended way to create C<Config::LotusNotes::Configuration> 
objects is to use the
L<default_configuration()|Config::LotusNotes/item_default_configuration> and 
L<all_configurations()|Config::LotusNotes/item_default_configuration>
methods of L<Config::LotusNotes|Config::LotusNotes>.

=item get_environment_value($item_name);

Gets the value of the parameter named C<$item_name> from F<notes.ini>.
If there is no such parameter, C<undef> is returned.

In order to access values that were written by Lotus Notes via the environment 
functions, prefix the parameter name with "$".

=item set_environment_value($item_name, $value);

Writes a parameter/value pair to F<notes.ini>.
If the entry exists, it will be updated with the new value.
If the value is C<undef>, the whole entry is removed.

If you want the parameter to be accessible to Lotus Notes via the environment 
functions, prefix its name with "$".

If you write to a F<notes.ini> file with this function, its entries will 
be saved in random order.

=back

=head1 BUGS AND LIMITATIONS

See L<Config::LotusNotes/"BUGS AND LIMITATIONS">.

=head1 EXAMPLES

See L<Config::LotusNotes/EXAMPLES>.

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2006 HS - Hamburger Software GmbH & Co. KG.
All rights reserved.

This module is free software; you can redistribute it and/or modify it 
under the same terms as Perl itself.

This library is distributed in the hope that it will be useful, but 
without any warranty; without even the implied warranty of 
merchantibility or fitness for a particular purpose.

=head1 AUTOR

Harald Albers, albers@cpan.org

See the F<Changes> file for the change history.

=cut


lib/Config/LotusNotes.pm
package Config::LotusNotes;
use strict;
#use warnings;
use Carp;
use Win32::TieRegistry Delimiter => '/';
use Config::LotusNotes::Configuration;

our $VERSION = '0.33';

# constructor ----------------------------------------------------------------

sub new {
    my ($classname, %options) = @_;
    return bless {%options}, $classname;
}


# public methods -------------------------------------------------------------

sub default_configuration {
    my ($self) = @_;
    my $path = _normalize_path($self->_get_default_location());
    return Config::LotusNotes::Configuration->new(path => $path);
}


sub all_configurations {
    my ($self) = @_;
    my @configurations;
    foreach my $path ($self->_get_all_locations()) {
        my $config;
        # skip invalid installs.
        print STDERR "-- Verifying $path - "  if $self->{debug};
        if ( eval {$config = Config::LotusNotes::Configuration->new(path => $path)} ) {
            print STDERR "OK\n"  if $self->{debug};
            push @configurations, $config;
        }
        else {
            my $errmsg = $@; 
            print STDERR "NOK: $errmsg\n"  if $self->{debug};
            carp $errmsg if $errmsg =~ /^Error parsing /;
        }
    }
    print STDERR "-- Returning " . @configurations . " configuration objects\n"  if $self->{debug};
    return @configurations;
}


# private methods ------------------------------------------------------------

sub _get_default_location {
    my ($self) = @_;
    print STDERR "-- Searching default install\n"  if $self->{debug};

    # first try the handler for the Notes class.
    if (my $class_handler = $self->_get_notes_handler()) {
        return $class_handler;
    }

    # if unsuccessful, try default product keys for Notes and Domino.
    foreach my $product qw(Notes Domino) {
        print STDERR "--  Searching default $product product key\n"  if $self->{debug};
        if (my $path = $Registry->{"LMachine/SOFTWARE/Lotus/$product/Path"}) {
            print STDERR "--   Found $path\n"  if $self->{debug};
            return _normalize_path($path);
        }
    }
    croak 'No Lotus Notes/Domino installation found';
}


sub _get_all_locations {
    my ($self) = @_;
    print STDERR "-- Searching for all installs\n"  if $self->{debug};
    my @all_paths = (
        $self->_get_notes_handler(), 
        $self->_get_registered_locations(), 
        $self->_get_shared_libs_locations(), 
        $self->_get_typelib_locations(),
    );
    # remove duplicates
    my (%seen, @result);
    print STDERR "-- Removing duplicates\n"  if $self->{debug};
    foreach my $path (@all_paths) {
        push @result, $path unless $seen{$path}++;
    }
    return @result;
}


sub _get_notes_handler {
    my ($self) = @_;
    # one install is registered for the Notes class.
    print STDERR "--  Searching Notes class handler\n"  if $self->{debug};
    if (my $path = $Registry->{"Classes/Notes/Shell/Open/Command//"}) {
        $path =~ s/^"([^"]+)".*/$1/;  # first quoted element
        $path =~ s/[^\\]+$//;  # shrink to path
        print STDERR "--   Found $path\n"  if $self->{debug};
        return _normalize_path($path);
    }
    return;
}


sub _get_registered_locations {
# finds installs in version specific keys under $RegNotesRoot
    my ($self) = @_;
    my @result;
    foreach my $product qw(Notes Domino) {
        print STDERR "--  Searching default $product product key\n"  if $self->{debug};
        my $product_root = $Registry->{"LMachine/SOFTWARE/Lotus/$product/"};
        if (my $path = $product_root->{'Path'}) {
            print STDERR "--   Found $path\n"  if $self->{debug};
            push @result, _normalize_path($path);
        }
        
        # search for version specific keys
        foreach my $key (keys %$product_root) {
            if ($key =~ m|^\d+(\.\d+)?/$|) {
                if ($self->{debug}) {
                    (my $version = $key) =~ s|/$||;
                    print STDERR "--  Searching $product $version product key\n";
                }
                if (my $path = $product_root->{$key . 'Path'}) {
                    print STDERR "--   Found $path\n"  if $self->{debug};
                    push @result, _normalize_path($path);
                }
            }
        }
    }
    return @result;
}


sub _get_shared_libs_locations {
# search for shared dll entries
    my ($self) = @_;
    print STDERR "--  Searching shared dll entries\n"  if $self->{debug};
    my @result;

    my $shared_dlls = $Registry->{'LMachine/SOFTWARE/Microsoft/Windows/CurrentVersion/SharedDLLs/'};
    foreach my $path (keys %$shared_dlls) {
        if ($path =~ /nlsxbe.dll$/i) {
            $path =~ s|^/||;
            $path =~ s/[^\\]*$//;  # remove program
            print STDERR "--   Found $path\n"  if $self->{debug};
            push @result, _normalize_path($path);
        }
    }
    return @result;
}


sub _get_typelib_locations {
# search the Lotus.NotesSession COM server's typelibs for installs
    my ($self) = @_;

    print STDERR "--  Searching the Lotus.NotesSession COM server's typelibs\n"  if $self->{debug};
    my @result;

    # get the COM Server's CLSID
    my $com_clsid = $Registry->{'Classes/Lotus.NotesSession/CLSID//'}
        or return;

    # get the CLSID of its typelib key
    my $typelib_clsid = $Registry->{"Classes/CLSID/$com_clsid/TypeLib//"}
        or return;

    # search all available typelib versions for paths
    my $typelib = $Registry->{"Classes/TypeLib/$typelib_clsid/"};
    foreach my $key (keys %$typelib) {
        if ($self->{debug}) {
            (my $version = $key) =~ s|/$||;
            print STDERR "--   Searching typelib version $version\n";
        }
        if (my $path = $typelib->{$key . '0/win32//'}) {
            $path =~ s/[^\\]*$//;  # remove program
            print STDERR "--    Found $path\n"  if $self->{debug};
            push @result, _normalize_path($path);
        }
        if (my $path = $typelib->{$key . 'HELPDIR//'}) {
            print STDERR "--    Found $path\n"  if $self->{debug};
            push @result, _normalize_path($path);
        }
    }
    return @result;
}


# utility class methods ------------------------------------------------------

sub _normalize_path {
    # append backslash, lowercase
    my ($path) = @_;
    $path .= '\\' unless $path =~ /\\$/;
    return lc $path;
}


1;


=head1 NAME

Config::LotusNotes - Access Lotus Notes/Domino configuration

=head1 VERSION

This documentation refers to C<Config::LotusNotes> 0.33, 
released Feb 7, 2011.

=head1 SYNOPSIS

  $factory = Config::LotusNotes->new();

  # access default installation
  $conf = $factory->default_configuration();
  $data = $conf->get_environment_value('Directory');
  $conf->set_environment_value('$NotesEnvParameter', 'value');

  # find all installations 
  @installs = $factory->get_all_configurations();

=head1 DESCRIPTION

C<Config::LotusNotes> gives you a view of your local Lotus Notes/Domino 
installations from the filesystem's perspective.
Its main purpose is to read and manipulate the Notes configuration file, 
F<notes.ini>. 

The module can handle multiple installations.

You can use it to

 - enumerate local Notes/Domino installations
 - gather basic information about your local Notes/Domino installations 
 - exchange data with Lotus Notes via the environment functions.

A C<Config::LotusNotes> object searches the Windows registry for Lotus Notes
installations, which can then be accessed in their representations as 
L<Config::LotusNotes::Configuration|Config::LotusNotes::Configuration> objects.

=head2 The Lotus Notes environment

The Lotus Notes environment is often used to store local user preferences 
and to share information between separate parts of an application. 

The Lotus Notes formula language has the C<@Environment> and C<@SetEnvironment>
functions and the C<ENVIRONMENT> keyword to access the program's environment.
Lotus script uses the C<GetEnvironmentValue>, C<GetEnvironmentString>,
C<SetEnvironmentVar> and C<Environ> functions for that purpose.
The Lotus Notes environment is stored in the F<notes.ini> file, which is 
instantly updated after each change to the environment. 
This allows you to communicate data to external programs.

Unfortunately, Lotus Notes does not recognize external changes to 
F<notes.ini> while it is running. 
If you need to push data to a running instance of Lotus Notes, you can use the 
environment functions of the corresponding OLE object as shown in L<SEE ALSO>.
There might be problems with simultaneous programmatic and user access 
to the same Lotus Notes session.  

=head1 METHODS

=over 4

=item new();

Constructor, returns a C<Config::LotusNotes> object that can give you
L<Config::LotusNotes::Configuration|Config::LotusNotes::Configuration> objects
via its default_configuration() and all_configurations() methods.

=item default_configuration();

Returns a L<Config::LotusNotes::Configuration|Config::LotusNotes::Configuration> 
object for the default Lotus Notes installation.
The default installation is the one that is registered in the Windows registry 
as the handler for the C<Notes> class.

If there is only one version of Lotus Notes installed on your machine, 
this is what you want.

This method will throw an 'No Lotus Notes/Domino installation found' exception 
if it cannot find any Lotus Notes installation.

=item all_configurations();

This gives you an array containing one 
L<Config::LotusNotes::Configuration|Config::LotusNotes::Configuration> object 
for each Lotus Notes/Domino installation found on your machine.
If no installation is found, an empty array is returned.

=back

=head1 SEE ALSO

An alternative way of accessing Lotus Notes/Domino is via its OLE and COM 
scripting capabilities. Here is an OLE example:

  use Win32::OLE;

  # print Lotus Notes version:
  $Notes = Win32::OLE->new('Notes.NotesSession')
      or die "Cannot start Notes.NotesSession object.\n";
  ($Version) = ($Notes->{NotesVersion} =~ /\s*(.*\S)\s*$/); # remove blanks
  printf "Running Notes %s on %s.\n", $Version, $Notes->Platform;

  # write value to environment
  print "Setting $key to $value\n";
  $session->SetEnvironmentVar('$NotesEnvParameter', 'test value');

This will start an instance of Lotus Notes if none is already running.
See the Lotus Notes designer documentation for more information.

=head1 DIAGNOSTICS

Call the constructor method new() with the option C<debug =E<gt> 1> to get 
diagnostic information on the search progress.

=head1 DEPENDENCIES

This module only works with Microsoft Windows.
It has been tested on Windows NT, 2000, XP and 7 (64-bit).
It uses L<Win32::TieRegistry|Win32::TieRegistry>,
L<Config::IniHash|Config::IniHash> and L<File::HomeDir|File::HomeDir>
(the latter two not being standard modules).
The test require Test::More.
Optional modules for the tests are Test::Pod and Test::Pod::Coverage.

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests through the web interface at 
http://rt.cpan.org/Public/Bug/Report.html?Queue=Config-LotusNotes 

=head2 Problems locating installations

Lotus Notes/Domino stores information about the installed versions in 
registry keys that are specific  to the main version number only, 
e.g. 5.0, 6.0 and 7.0, with ".0" being fix.
Each additional installation will overwrite the data of any previous 
installation that has the same main version number.

This module works around this problem by searching several additional places 
in the registry for possible installation locations. 
In complex installations this might not find all instances.

Please bear in mind that such complex situations can only be created if you 
cheat the Notes installer by renaming the paths of your existing installations 
before each additional installation.
The normal behaviour of the installer is to force you to update your previous 
installation. 
So in real life, there should be no problem with missed installations.   

=head2 Malformed notes.ini files

Under certains conditions a F<notes.ini> file may contain malformed lines,
i.e. lines that do not match the pattern C<parameter=value>.

Such lines are ignored when reading values. 
If you write back values to a corrupt F<notes.ini> file using the
L<set_environment_value()|Config::LotusNotes::Configuration/set_environment_value> 
function, the malformed lines are removed from F<notes.ini>.

If a F<notes.ini> file cannot be parsed at all, a warning will be issued and the 
corresponding installation will be skipped by all_configurations()Z<>. 
default_configuration() will throw an "Error parsing ..." exception in that case.

Malformed F<notes.ini> files can be reproduced by writing multiline values to the
environment, e.g. with Notes formula code like this: 
C<@SetEnvironment("testvalue"; "A"+@Char(10)+"B")>, which produces two lines, 
the second one just containing "B".
A successive read of testvalue will return just "A".

In my environment lines like

 3.2008 09:32:30

are quite common and seem to originate from the Smart Upgrade process.

=head2 Parameter order in notes.ini

If you write to a F<notes.ini> file with the
L<set_environment_value()|Config::LotusNotes::Configuration/set_environment_value> 
function, the entries in that file will appear in random order.
This should not pose any problems.

=head1 EXAMPLES

=head2 code example

  use Config::LotusNotes;
  $factory = Config::LotusNotes->new();

  # get default LotusNotes installation
  $conf = $factory->default_configuration();
  print 'Lotus Notes ', $conf->version, ' installed in ', $conf->notespath, "\n";

  # retrieving and setting notes.ini values
  # get name of the user's mail file.
  $mail_file = $conf->get_environment_value('MailFile');
  # store a value in notes.ini
  $conf->set_environment_value('$NotesEnvParameter', 'test value');

  # find all installations 
  @all_confs = $factory->all_configurations();

  # print a table with version, type and path of all installations.
  # see demo\FindNotes.pl for an extended version.
  printf "%-8s %-7s %s\n", 'version', 'type', 'path';
  foreach my $conf (@all_confs) {
      printf "%-8s %-7s %s\n", 
          $conf->version, 
          $conf->is_server ? 'server' : 'client', 
          $conf->notespath;
  }

  # filter the list: only servers
  @servers = grep { $_->is_server } @all_confs;

=head2 demo scripts

This module also installs two simple demonstration scripts: 
F<FindNotes.pl> enumerates local Notes/Domino installations and
F<editnotesini.pl> reads and changes notes.ini parameters.

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2006 HS - Hamburger Software GmbH & Co. KG.
All rights reserved.

This module is free software; you can redistribute it and/or modify it 
under the same terms as Perl itself.

This library is distributed in the hope that it will be useful, but 
without any warranty; without even the implied warranty of 
merchantibility or fitness for a particular purpose.

=head1 AUTOR

Harald Albers, albers@cpan.org

See the F<Changes> file for the change history.

=cut
